[patch] Fix the script in suite_db2_setup_for_manage role to make changes via db2ucluster CR

csschen opened the pull request:
For story #607 , fix the script issue for DB2 setup. Details in [issuecomment-1637508655](https://github.com/ibm-mas/ansible-devops/issues/607#issuecomment-1637508655)

durera reviewed (commented):
Do you have links to logs where this has been tested?

csschen commented:
> Do you have links to logs where this has been tested?

@durera The Ansible role run on fyre machine, 

For ROKS, the target cluster is **masms-devops-1-roks-np-wdc** on **2116608 - MAS DevOps** account.
For ROSA, the target cluster is **ss-04-rs-use1** on **massaasstg-ibm-aws** account.

Below are the background and testing details:

From arlindo@us.ibm.com's request, script setupdb should not do any db2 cfg settings, as those should belong into the db2u cluster CR.

DBA [leqingfu@cn.ibm.com](mailto:leqingfu@cn.ibm.com) help reviewed the DB2 setup script, the suggestion is that all the db2 config, db2 dbm config, db2 set command should not execute on DB2 instance directly, instead of changing the DB2uCluster CR. Those contents should removed from the script.
He also help refined and fixed the marker file missing issue, will query the results from DB2 instance per time before create the DB2 objects.

The new script tested successfully by Ansible role [suite_db2_setup_for_manage](https://github.com/ibm-mas/ansible-devops/tree/db2-prepare-maxinst/ibm/mas_devops/roles/suite_db2_setup_for_manage), below are the db2 setup script test logs:

```
"stderr": "Defaulted container \"db2u\" out of: db2u, init-labels (init), init-kernel (init)",
    "stderr_lines": [
        "Defaulted container \"db2u\" out of: db2u, init-labels (init), init-kernel (init)"
    ],
    "stdout": "\n   Database Connection Information\n\n Database server        = DB2/LINUXX8664 11.5.8.0\n SQL authorization ID   = DB2INST1\n Local database alias   = BLUDB\n\nCREATE BUFFERPOOL MAXBUFPOOL SIZE 8192 AUTOMATIC PAGESIZE 32 K\nDB20000I  The SQL command completed successfully.\n\nCREATE BUFFERPOOL MAXBUFPOOLINDX SIZE 8192 AUTOMATIC PAGESIZE 32 K\nDB20000I  The SQL command completed successfully.\n\nCREATE BUFFERPOOL MAXTEMPBP SIZE 8192 AUTOMATIC PAGESIZE 32 K\nDB20000I  The SQL command completed successfully.\n\nCREATE TABLESPACE MAXDATA PAGESIZE 32 K MANAGED BY AUTOMATIC STORAGE BUFFERPOOL MAXBUFPOOL\nDB20000I  The SQL command completed successfully.\n\nCREATE TABLESPACE MAXINDEX PAGESIZE 32 K MANAGED BY AUTOMATIC STORAGE BUFFERPOOL MAXBUFPOOLINDX\nDB20000I  The SQL command completed successfully.\n\nCREATE TEMPORARY TABLESPACE MAXTEMP PAGESIZE 32 K MANAGED BY AUTOMATIC STORAGE BUFFERPOOL MAXTEMPBP\nDB20000I  The SQL command completed successfully.",
    "stdout_lines": [
        "",
        "   Database Connection Information",
        "",
        " Database server        = DB2/LINUXX8664 11.5.8.0",
        " SQL authorization ID   = DB2INST1",
        " Local database alias   = BLUDB",
        "",
        "CREATE BUFFERPOOL MAXBUFPOOL SIZE 8192 AUTOMATIC PAGESIZE 32 K",
        "DB20000I  The SQL command completed successfully.",
        "",
        "CREATE BUFFERPOOL MAXBUFPOOLINDX SIZE 8192 AUTOMATIC PAGESIZE 32 K",
        "DB20000I  The SQL command completed successfully.",
        "",
        "CREATE BUFFERPOOL MAXTEMPBP SIZE 8192 AUTOMATIC PAGESIZE 32 K",
        "DB20000I  The SQL command completed successfully.",
        "",
        "CREATE TABLESPACE MAXDATA PAGESIZE 32 K MANAGED BY AUTOMATIC STORAGE BUFFERPOOL MAXBUFPOOL",
        "DB20000I  The SQL command completed successfully.",
        "",
        "CREATE TABLESPACE MAXINDEX PAGESIZE 32 K MANAGED BY AUTOMATIC STORAGE BUFFERPOOL MAXBUFPOOLINDX",
        "DB20000I  The SQL command completed successfully.",
        "",
        "CREATE TEMPORARY TABLESPACE MAXTEMP PAGESIZE 32 K MANAGED BY AUTOMATIC STORAGE BUFFERPOOL MAXTEMPBP",
        "DB20000I  The SQL command completed successfully."
    ]
```

The DBA leqingfu@cn.ibm.com help confirmed above log is working as expected.
The script tested on both ROKS and ROSA.
Below are the Ansible role test logs.

[roks_test.log](https://github.com/ibm-mas/ansible-devops/files/12066866/roks_test.log)
[rosa_test.log](https://github.com/ibm-mas/ansible-devops/files/12066867/rosa_test.log)

whitfiea commented:
@csschen can you resolve the merge conflicts please.

csschen commented:
@whitfiea Done. All the db2 config, db2 dbm config, db2 set command should not execute on DB2 instance directly, instead of changing the DB2uCluster CR.

csschen commented:
@whitfiea @CAIOFCP As we discussed, merged the DB2 parameters reviewed by DBA into default variables on `ibm.mas_devops.db2` role.

Below is the Ansible role test log.
[db2_test.log](https://github.com/ibm-mas/ansible-devops/files/12333230/db2_test.log)

Noted that those parameters only configured on DB2ucluster CR, below script execution is required to make the changes take effect on DB2 instance, will cause downtime.
`/db2u/scripts/apply-db2cfg-settings.sh`

wangqymro reviewed (commented):
Recommend logfilesize set to 64k, and primary set to 150, and secondary to 100. This only mount to 16G disk space.

leqingfu commented:
> Recommend logfilesize set to 64k, and primary set to 150, and secondary to 100. This only mount to 16G disk space.

LOGFILSIZ 32768 LOGPRIMARY 100 LOGSECOND 156 apply to most of our clients' business, and are running on SaaS Flex for years. I would not recommend to change them unless we have strong reason.

wangqymro commented:
What I mentioned is what performance team’s recommendation
@Brian ***@***.***> chime in place.
Good to know that 32k is working and its been fine.
If it has been fine, we should certainly use that value.
Internal testing of our dev pipeline with a smaller log size (16k and less files) show problem when some industry solutions are installed. So, we want to move up to be sufficient if no risk of other concerns.

Jenny

From: leqingfu ***@***.***>
Sent: Wednesday, August 16, 2023 3:01 AM
To: ibm-mas/ansible-devops ***@***.***>
Cc: Jenny Wang ***@***.***>; Review requested ***@***.***>
Subject: [EXTERNAL] Re: [ibm-mas/ansible-devops] [patch] Fix the script to setupdb for DB2 (PR #917)

Recommend logfilesize set to 64k, and primary set to 150, and secondary to 100. This only mount to 16G disk space. LOGFILSIZ 32768 LOGPRIMARY 100 LOGSECOND 156 apply to most of our clients' business, and are running on SaaS Flex for years. 
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.
    Report Suspicious  <https://us-phishalarm-ewt.proofpoint.com/EWT/v1/PjiDSg!2O-q6PXc4ROy2c-QEOpvxTUo2gqdsWqYeI9rNtct_WfLoARopsd0s6MVG6Md6w1u2CgiqLQ57Ra7Yq_P2E3OUp8_YiIK9F0kCXGeXbnxjQ$>   ‌
ZjQcmQRYFpfptBannerEnd

Recommend logfilesize set to 64k, and primary set to 150, and secondary to 100. This only mount to 16G disk space.

LOGFILSIZ 32768 LOGPRIMARY 100 LOGSECOND 156 apply to most of our clients' business, and are running on SaaS Flex for years. I would not recommend to change them unless we have strong reason.

—
Reply to this email directly, view it on GitHub<https://github.com/ibm-mas/ansible-devops/pull/917#issuecomment-1680069697>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AELEYDRS6KR3UJN5GAPHRD3XVRV3LANCNFSM6AAAAAA2MQ42KU>.
You are receiving this because your review was requested.Message ID: ***@***.***>

leqingfu commented:
> What I mentioned is what performance team’s recommendation @brian ***@***.***> chime in place. Good to know that 32k is working and its been fine. If it has been fine, we should certainly use that value. Internal testing of our dev pipeline with a smaller log size (16k and less files) show problem when some industry solutions are installed. So, we want to move up to be sufficient if no risk of other concerns. Jenny From: leqingfu ***@***.***> Sent: Wednesday, August 16, 2023 3:01 AM To: ibm-mas/ansible-devops ***@***.***> Cc: Jenny Wang ***@***.***>; Review requested ***@***.***> Subject: [EXTERNAL] Re: [ibm-mas/ansible-devops] [patch] Fix the script to setupdb for DB2 (PR #917) Recommend logfilesize set to 64k, and primary set to 150, and secondary to 100. This only mount to 16G disk space. LOGFILSIZ 32768 LOGPRIMARY 100 LOGSECOND 156 apply to most of our clients' business, and are running on SaaS Flex for years.  ZjQcmQRYFpfptBannerStart This Message Is From an External Sender This message came from outside your organization. Report Suspicious <[https://us-phishalarm-ewt.proofpoint.com/EWT/v1/PjiDSg!2O-q6PXc4ROy2c-QEOpvxTUo2gqdsWqYeI9rNtct_WfLoARopsd0s6MVG6Md6w1u2CgiqLQ57Ra7Yq_P2E3OUp8_YiIK9F0kCXGeXbnxjQ$](https://us-phishalarm-ewt.proofpoint.com/EWT/v1/PjiDSg!2O-q6PXc4ROy2c-QEOpvxTUo2gqdsWqYeI9rNtct_WfLoARopsd0s6MVG6Md6w1u2CgiqLQ57Ra7Yq_P2E3OUp8_YiIK9F0kCXGeXbnxjQ%24)> ‌ ZjQcmQRYFpfptBannerEnd Recommend logfilesize set to 64k, and primary set to 150, and secondary to 100. This only mount to 16G disk space. LOGFILSIZ 32768 LOGPRIMARY 100 LOGSECOND 156 apply to most of our clients' business, and are running on SaaS Flex for years. I would not recommend to change them unless we have strong reason. — Reply to this email directly, view it on GitHub<[#917 (comment)](https://github.com/ibm-mas/ansible-devops/pull/917#issuecomment-1680069697)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AELEYDRS6KR3UJN5GAPHRD3XVRV3LANCNFSM6AAAAAA2MQ42KU>. You are receiving this because your review was requested.Message ID: ***@***.***>

Thanks for the information. 
Our databases are numerous and vary widely in capacity. I don't think there is the only criteria for all of our databases. But I think our current settings are a good start, they have been well tested, and apply to most of clients.

whitfiea commented:
@csschen I believe this change needs a corresponding change in the cli repo so that new parameters are passed into the db2 task in the pipeline

csschen commented:
@whitfiea There are more than 50 DB2 parameters required to apply, I'm not sure if mas cli can consume those as input parameters. If there are already default values for db2config, db2config, registry on `ibm.mas_devops.db2` role, why don't we use existing variables?

csschen commented:
@CAIOFCP As we discussed, added the logic to execute the script `/db2u/scripts/apply-db2cfg-settings.sh` to make db2 config changes take effect. It will cause DB2 instance downtime. 

This controlled by flag `db2_enforce_config`, the default value set to False.
If set to True, the role `ibm.mas_devops.db2` should be executed during customer maintenance window or for new environment.

Below is the test log.
[db2_test.log](https://github.com/ibm-mas/ansible-devops/files/12406858/db2_test.log)

whitfiea reviewed (changes requested):
@csschen The default I believe should be True to apply the config. There a few issues that need to be confirmed before we can merge:
- A user running these roles before this PR will expect the same outcome with this PR i.e. DB2 configured and working as expected or has the behaviour changed?
- The DB2 configuration has moved to the db2ucluster CR that is now applied to anyone running the DB2 role. Previously these setting were only when the `suite_db2_setup_for_manage` role was run. Are these new settings on the db2ucluster be correct and work for users that install DB2 for Monitor or IoT? If not then we need to make sure those users are not impacted.

csschen commented:
> @csschen The default I believe should be True to apply the config. There a few issues that need to be confirmed before we can merge:
> 
> * A user running these roles before this PR will expect the same outcome with this PR i.e. DB2 configured and working as expected or has the behaviour changed?
> * The DB2 configuration has moved to the db2ucluster CR that is now applied to anyone running the DB2 role. Previously these setting were only when the `suite_db2_setup_for_manage` role was run. Are these new settings on the db2ucluster be correct and work for users that install DB2 for Monitor or IoT? If not then we need to make sure those users are not impacted.

@whitfiea Changed the default value to true for `db2_enforce_config` flag, that means the DB2 instance will be restart by default.
Discussed with DBA, all the changes has nothing to do with application, but to DB2 itself, those changes are worked if it's DB2. 
After merge this PR, the DB2 configs will be enhanced and performance will be improved.
FYI @leqingfu

csschen commented:
As we discussed on slack thread https://ibm-watson-iot.slack.com/archives/C040TETRV2T/p1693922018789469, to ensure the PR change not impact iot/monitor, @andrercm will help do the FVT on MAS Dev account tmp cluster.
Please let me know if any questions. Thanks! @andrercm

andrercm commented:
@csschen is this ready to be tested for iot/monitor/predict databases?

csschen commented:
@andrercm Just finished the testing for DB2 role on MAS provisioning automation pipeline, unfortunately we can't set the db2 config parameters on the default value when db2 initializing, after restart the DB2, then caused the jdbc connection issue for MAS core, broken the pipeline.

Below is the jdbc connection issue:
```
JDBC configuration was unable to be verified: [IBM][CLI Driver] SQL1116N  A connection to or activation of database "BLUDB" failed because the database is in BACKUP PENDING state.  SQLSTATE=57019 SQLCODE=-1116
```

Discussed with DBA @leqingfu , the problem should gone after execute below command:
```
db2 backup db bludb to /mnt/backup
```
But this is not the safety way to make the DB2 instance working as expected.
The solution is that initialize the DB2 with default value on db2 role first, then apply the DB2 config parameters, restart the DB2 instance. Details in [issuecomment-63496300](https://github.ibm.com/maximoappsuite/tracker-masms/issues/3175#issuecomment-63496300)

Below are the steps for testing:

1. Installed MAS core, Manage, DB2 with default value, this is part of provisioning pipeline
2. Applied the DB2 config parameters by role [ibm.sre_devops.db2_dbonfig](https://github.ibm.com/sreaiapps/sre-ansible-devops/tree/db2config-saas/ansible/ibm/sre_devops/roles/db2_dbonfig), this is part of provisioning pipeline
3. Re-create the JDBC config CR on MAS core
4. The JDBC config CR becomes Ready, Running, Successful

The DB2 role on this PR will be rollbacked, the changes regarding DB2 parameters configs/settings will be handled by role [ibm.sre_devops.db2_dbonfig](https://github.ibm.com/sreaiapps/sre-ansible-devops/tree/db2config-saas/ansible/ibm/sre_devops/roles/db2_dbonfig).

We can consider merge into public collection if need. @CAIOFCP @durera @andrercm

whitfiea commented:
@csschen The role https://github.ibm.com/sreaiapps/sre-ansible-devops/tree/db2config-saas/ansible/ibm/sre_devops/roles/db2_dbonfig is not something a customer can run. If we merge these changes in this PR today will a customer that doesn't have access to the db2_dbconfig role be impacted by these changes? My feels is yes based on what you have said.

csschen commented:
@whitfiea I've opened the issue https://github.com/ibm-mas/ansible-devops/issues/1020 to migrate the Ansible role [ibm.sre_devops.db2_dbonfig](https://github.ibm.com/sreaiapps/sre-ansible-devops/tree/db2config-saas/ansible/ibm/sre_devops/roles/db2_dbonfig) to public collection so that customer can benefit the automation.

The current change for this PR removed the db2 configs/settings, those changes should applied on DB2cluster CR, it will be lost even it's applied on DB2 instance directly once the DB2 instance pod re-created.

The remaining is create the DB2 object on the right way.

If merged the PR, the db2 configs/settings required to applied.
If not merge, there is potential risk to loss the db2 configs/settings if DB2 instance pod re-created.

My suggestion is merge this PR with the issue https://github.com/ibm-mas/ansible-devops/issues/1020 completed.

whitfiea commented:
The current state of this PR is that the change reported in issue https://github.com/ibm-mas/ansible-devops/issues/1020 needs to be completed before this PR can be merged. We can not allow the db2 and db2_setup_for_manage role to not function as expected (i.e. setup Manage with DB2) without needing to run an internal role within IBM. If the function in this internal role needs to be present then it must be included in this change.

Waiting on the SRE team to work on https://github.com/ibm-mas/ansible-devops/issues/1020

csschen commented:
@whitfiea The Ansible role [ibm.sre_devops.db2_dbonfig](https://github.ibm.com/sreaiapps/sre-ansible-devops/tree/db2config-saas/ansible/ibm/sre_devops/roles/db2_dbonfig) migrated to public Ansible collection, this role should run on existing or newly created DB2 instance to enhance the DB2 configs/settings.

Details in [issuecomment-1733186351](https://github.com/ibm-mas/ansible-devops/issues/1020#issuecomment-1733186351)

This PR is ready for review now.

whitfiea reviewed (changes requested):
Would it not make more sense to incorporate this new role into the existing suite_db2_setup_for_manage role? As that is what this new role is doing? or is there a reason why it has to be a new role?

The default values (namespace, dbname etc) should be the same as before and anything related to "sre" should be removed. Also you can't assume the db name is bludb which I saw in some calls.

csschen commented:
> Would it not make more sense to incorporate this new role into the existing suite_db2_setup_for_manage role? As that is what this new role is doing? or is there a reason why it has to be a new role?
> 
> The default values (namespace, dbname etc) should be the same as before and anything related to "sre" should be removed. Also you can't assume the db name is bludb which I saw in some calls.

Below are the reasons to create new role:

- The new role `db2_dbconfig` not only covers newly created environment, but also for existing environment. The role `suite_db2_setup_for_manage` should be part of the DB2 initialization. Merging them would be a little confuse.
- Both `suite_db2_setup_for_manage` and `db2_dbconfig` are part of provisioning pipeline, the former executed after DB2 instance initialized, the latter executed after MAS core/Manage are ready. This already tested and proved worked. Not sure the behavior for DB2 if merging them into one role. This also caused the change for current SRE automation pipeline if sync with the latest version.
- We can safe add the role `db2_dbconfig` on any other automation pipeline, it can also be executed separately for existing environments.

Besides that, changed the default value to keep sync with db2 role, also removed anything related "sre", the role didn't use dbname.

csschen commented:
@whitfiea As we discussed on slack thread https://ibm-watson-iot.slack.com/archives/C040TETRV2T/p1695717231984079?thread_ts=1693922018.789469&cid=C040TETRV2T,
Merged the db2 config logic into `suite_db2_setup_for_manage` so that customer can easily pick up the change in their pipeline, also no change required on the cli.

The merged role `suite_db2_setup_for_manage` working as expected.

Below is the test log.
[db2_setup_config.log](https://github.com/ibm-mas/ansible-devops/files/12734211/db2_setup_config.log)

whitfiea commented on ibm/mas_devops/roles/suite_db2_setup_for_manage/defaults/main.yml:
```diff
@@ -1,13 +1,7 @@
-# templateable in automation such as this :(
-db2_schema: "{{ lookup('env', 'MAS_APP_SETTINGS_DB2_SCHEMA') | default('maximo', true)}}"
+# Variables for db2_dbconfig
+db2_config_version: "{{ lookup('env', 'DB2_CONFIG_VERSION') | default('', true) }}"
```
would this not default to 1.0.0?

csschen replied:
Changed default value to 1.0.0

csschen commented on ibm/mas_devops/roles/suite_db2_setup_for_manage/tasks/apply-db2-config-settings.yml:
```diff
@@ -0,0 +1,53 @@
+  - name: Check DB2 cfg is take effect
+    shell: |
+      oc exec -n {{db2_namespace}} {{db2_pod_name}} -- su -lc '/db2u/scripts/apply-db2cfg-settings.sh | tee /tmp/apply-db2cfg-settings.log' db2inst1
+      oc exec -n {{db2_namespace}} {{db2_pod_name}} -- su -lc 'db2 get db cfg for bludb | grep "(CHNGPGS_THRESH) = 40"' db2inst1
```
changed the bludb to dbname variable.

csschen commented on ibm/mas_devops/roles/suite_db2_setup_for_manage/README.md:
```diff
@@ -56,6 +56,19 @@ The size of the tablespace indexes in the database.
+
+- **Required**
+- Environment Variable: `DB2_CONFIG_VERSION`
+- Default: None
```
Changed default value to 1.0.0

csschen commented on ibm/mas_devops/roles/db2_dbonfig/README.md:
```diff
@@ -0,0 +1,64 @@
+db2_dbconfig
```
Removed.

csschen commented on ibm/mas_devops/roles/suite_db2_setup_for_manage/tasks/apply-db2-config-settings.yml:
```diff
@@ -0,0 +1,53 @@
+
+  when:
+    - db2_namespace is defined
+    - db2_namespace != "mas-cpd"
```
mas-cpd means the DB2 hosted on CP4D, we don't need this check for CP4D, I can remove this if this role only applied for DB2 standalone.

whitfiea replied:
Yes it only applies to db2u standalone, so I think we can remove it

csschen replied:
Removed the mas-cpd check.

whitfiea reviewed (changes requested):
@csschen I think remove that check for mas-cpd and we should be ok to merge this

csschen commented:
@whitfiea Removed the `mas-cpd` check, now all the conversions has been resolved, please help review and merge. Thanks!
